### ISSUE TITLE
Ignore negation doesn't work on directories

### ISSUE THREAD

**Issue by Gilnaa** [Gilnaa/globwalk#5]
To reproduce:
```
mkdir test_area
touch test_area/{.,Pictures}/{a,b,c}.{jpg,png,gif}
```

Code:
```
fn main() {
    let walker = globwalk::GlobWalker::from_patterns("/path/to/test_area", &["*.{png,jpg,gif}", "!Pictures"])
        .unwrap()
        .into_iter()
        .filter_map(Result::ok);
    for img in walker {
        println!("{:?}", img.path());
    }
}
```

Result:
```
"/home/gilnaa/proj/ttt/tree/a.gif"
"/home/gilnaa/proj/ttt/tree/a.jpg"
"/home/gilnaa/proj/ttt/tree/a.png"
"/home/gilnaa/proj/ttt/tree/b.gif"
"/home/gilnaa/proj/ttt/tree/b.jpg"
"/home/gilnaa/proj/ttt/tree/b.png"
"/home/gilnaa/proj/ttt/tree/c.gif"
"/home/gilnaa/proj/ttt/tree/c.jpg"
"/home/gilnaa/proj/ttt/tree/c.png"
"/home/gilnaa/proj/ttt/tree/Pictures/a.gif"
"/home/gilnaa/proj/ttt/tree/Pictures/a.jpg"
"/home/gilnaa/proj/ttt/tree/Pictures/a.png"
"/home/gilnaa/proj/ttt/tree/Pictures/b.gif"
"/home/gilnaa/proj/ttt/tree/Pictures/b.jpg"
"/home/gilnaa/proj/ttt/tree/Pictures/b.png"
"/home/gilnaa/proj/ttt/tree/Pictures/c.gif"
"/home/gilnaa/proj/ttt/tree/Pictures/c.jpg"
"/home/gilnaa/proj/ttt/tree/Pictures/c.png"
```

**Comment by epage** [comment]
In cobalt, I built the logic myself
https://github.com/cobalt-org/cobalt.rs/blob/master/src/cobalt_model/files.rs#L120

I saw `ignore` now has a function to handle this and tried to use it
https://docs.rs/ignore/0.4.1/ignore/gitignore/struct.Gitignore.html#method.matched_path_or_any_parents

**Comment by Gilnaa** [comment]
We can always just revert to using globset and keep 2 seperate globsets, one positive and one negative.
It would solve both this and #10

**Comment by epage** [comment]
Looking at the implementation, I think its a bug in walkdir
https://docs.rs/ignore/0.4.1/src/ignore/gitignore.rs.html#216-240

They check the leaf for matches and then work to the root.  I think it is supposed to be the opposite, check the root and then work towards the leaf.

**Comment by Gilnaa** [comment]
Is this the expected behavior? @burntsushi

**Comment by BurntSushi** [comment]
I can't reproduce it using the `ignore` crate:

```rust
extern crate ignore;

use std::error;
use std::process;
use std::result;

use ignore::WalkBuilder;
use ignore::overrides::OverrideBuilder;

type Result<T> = result::Result<T, Box<error::Error + Send + Sync>>;

fn main() {
    if let Err(err) = try_main() {
        eprintln!("{}", err);
        process::exit(1);
    }
}

fn try_main() -> Result<()> {
    let overrides = OverrideBuilder::new("./test_area")
        .add("*.{png,jpg,gif}")?
        .add("!Pictures")?
        .build()?;
    let walker = WalkBuilder::new("./test_area")
        .overrides(overrides)
        .build();
    for result in walker {
        let dent = result?;
        println!("{}", dent.path().display());
    }
    Ok(())
}
```

Running this gives the expected output:

```
./test_area
./test_area/c.gif
./test_area/c.png
./test_area/c.jpg
./test_area/b.gif
./test_area/b.png
./test_area/b.jpg
./test_area/a.gif
./test_area/a.png
./test_area/a.jpg
```

Directory tree:

```
$ tree test_area/
test_area/
├── a.gif
├── a.jpg
├── a.png
├── b.gif
├── b.jpg
├── b.png
├── c.gif
├── c.jpg
├── c.png
└── Pictures
    ├── a.gif
    ├── a.jpg
    ├── a.png
    ├── b.gif
    ├── b.jpg
    ├── b.png
    ├── c.gif
    ├── c.jpg
    └── c.png
```

**Comment by epage** [comment]
Whats the difference between Overrides and Gitignore?

**Comment by Gilnaa** [comment]
I guess we can use WalkBuilder + standard_filters(false) + overrides

**Comment by epage** [comment]
Looking into it now.

**Comment by BurntSushi** [comment]
Overrides *is* a `Gitignore`:

```rust
pub struct Override(Gitignore);
```

It was built as a way to specify glob overrides on the command line, and it inverts the meaning of `!`. In a gitignore, a `!` means "whitelist," while in an override a `!` means "blacklist." You can see this in the definition of `num_ignores` and `num_whitelists` in the `Override` matcher:

```rust
    /// Returns the total number of ignore globs.
    pub fn num_ignores(&self) -> u64 {
        self.0.num_whitelists()
    }

    /// Returns the total number of whitelisted globs.
    pub fn num_whitelists(&self) -> u64 {
        self.0.num_ignores()
    }
```

Basically, overrides are a way of providing explicit globs to a walker, since you can't feed a `Gitignore` directly to a walker. (Instead, the walker reads gitignore files automatically if that's enabled.)

#10 also seems to work fine when using ignore's walker.

**Comment by Gilnaa** [comment]
Oh, great!

At one point we'll have to open the Andrew-Gallant-Facade WG.
Thanks :)

**Comment by BurntSushi** [comment]
:-) My suspicion is that you shouldn't be using `matched_path_or_any_parents` since you're walking the directory hierarchy. Instead, you should just use `matched`.

It may indeed be easier to just use the walker from `ignore`, but that will likely make optimizations harder. Although, `ignore` would probably want those same optimizations. The walker is, unfortunately, crazy complicated, because it supports a lot more stuff.

**Comment by epage** [comment]
Tried switching to ignore's `Walk` and ran into some problems as pointed out in #11.

**Comment by Gilnaa** [comment]
Let's avoid using walk, I'm trying to shoehorn overrides::Override instead of Gitignore

**Comment by epage** [comment]
I was originally going in that direction but thought I'd give `Walk` a try.  I'm assuming `Override` might give the same problems as #11.  We might want to stay at the `Gitignore::match` level so we can better customize what is returned.

**Comment by Gilnaa** [comment]
Nah, the problems in #11 stem from the fact that we're moving away from `walkdir` AFAICT.

Overrides has the same API as GitIgnore, and it fits pretty well, I just couldn't reproduce Andrew's results.

**Comment by BurntSushi** [comment]
What are you going to use instead of `walkdir`?

**Comment by Gilnaa** [comment]
I'm going to keep using `walkdir`

**Comment by epage** [comment]
I was exploring using ignore's wrapper around walkdir

**Comment by epage** [comment]
@Gilnaa I'll go ahead and hold off on any more investigation since we've narrowed down the approach atm and it sounds like you are already working on it.  Feel free to ping me if you run out of time and want me to play with it some more as well.

**Comment by Gilnaa** [comment]
All tests pass https://github.com/epage/globwalk/pull/1

**Comment by Gilnaa** [comment]
Closed by 30dd890f4a6ba5354c932dd71dc2cbbdda117091
